### PR TITLE
Fix for Issue #141

### PR DESCRIPTION
**Pull request by gebing**
Fix for Issue #141, Simultaneously invoke Java.use() will cause return wrong value.

### CODE REVIEW

**Comment by oleavr** [comment]
Thanks! I think this introduces another issue, where `Java.use()` might return two different wrappers for the same class name. So I think we want to put in a "placeholder" value as early as we can, and if another thread finds this it will have to wait until the real value is there or placeholder is gone. As we don't currently have a wait primitive, we can use a loop with `Thread.sleep()` as a temporary solution.

**Comment by gebing** [comment]
I think the case of `Java.use()` return two different wrappers will only rarely occur in the early stage of frida. Once `Java.use()` successfully returned, the following calls will use the wrapper from `classes[name]`. I think the only flaw is a little bit memory consumption and it will be free by garbage collection, and I think the "placeholder" solution will make the code to be a little bit complex.

**Comment by gebing** [comment]
If you still want the "placeholder", I can make a implementation and update the Pull Request.

**Comment by oleavr** [comment]
Yeah, I think we should ensure that it always does, to avoid surprises.

**Comment by gebing** [comment]
Sorry, i do not clearly understand what you mean. Keep the current or add "placeholder"?

**Comment by oleavr** [comment]
Sorry, I wasn't clear. Placeholder.

**Comment by gebing** [comment]
ok

**Comment by gebing** [comment]
@oleavr 
How about this pull request? When can release this patch?

File: lib/class-factory.js [hunk 1]
```diff
@@ -139,8 +140,27 @@ function ClassFactory (vm) {
...
+    return kclass;
+  }
+
+  function setUsedClass(className, kclass) {
+    if (kclass) {
+      classes[className] = kclass;
+    } else {
+      delete classes[className];
+    }
+  }
```
Comment (oleavr) [inline lib/class-factory.js]: Let's move these below where they're used. (Code should be readable from top to bottom, higher level details at the top with details further down.)

File: lib/class-factory.js [hunk 2]
```diff
@@ -139,8 +140,27 @@ function ClassFactory (vm) {
     }
   });
 
+  function getUsedClass(className) {
+    let kclass;
+    while ((kclass = classes[className]) === PENDING_USE) {
+      Thread.sleep(0.05);
+    }
+    if (!kclass) {
```
Comment (oleavr) [inline lib/class-factory.js]: Checks should be explicit:

```suggestion
    if (kclass === undefined) {
```

File: lib/class-factory.js [hunk 3]
```diff
@@ -139,8 +140,27 @@ function ClassFactory (vm) {
...
+      classes[className] = PENDING_USE;
+    }
+    return kclass;
+  }
+
+  function setUsedClass(className, kclass) {
+    if (kclass) {
```
Comment (oleavr) [inline lib/class-factory.js]:
```suggestion
    if (kclass !== undefined) {
```

File: lib/class-factory.js [hunk 4]
```diff
@@ -695,10 +723,7 @@ function ClassFactory (vm) {
   this.registerClass = registerClass;
 
   function ensureClass (getClassHandle, name) {
-    let klass = classes[name];
-    if (klass !== undefined) {
-      return klass;
-    }
+    let klass;
```
Comment (oleavr) [inline lib/class-factory.js]: This can be moved down to right before it's used.

File: lib/class-factory.js [hunk 5]
```diff
@@ -715,10 +740,15 @@ function ClassFactory (vm) {
         return superHandle;
       };
 
-      try {
-        superKlass = ensureClass(getSuperClassHandle, env.getClassName(superHandle));
-      } finally {
-        env.deleteLocalRef(superHandle);
+      const superClassName = env.getClassName(superHandle);
+      superKlass = getUsedClass(superClassName);
+      if (!superKlass) {
```
Comment (oleavr) [inline lib/class-factory.js]:
```suggestion
      if (superKlass === undefined) {
```

File: lib/class-factory.js (hunk 1 above)
Comment (gebing) [reply lib/class-factory.js]: Sorry, I can not find the right place for it. Do you mean only move `delete classes[className]` down, or move the whole function `setUsedClass()` down? 

Function  `setUsedClass()` is used by funtion `use()` and `ensureClass()`, so must be placed outside these two functions. And i also think the code of  `delete classes[className]` should not be move outside `setUsedClass()`.

Pls give me some suggestion.

Comment (oleavr) [reply lib/class-factory.js]: I mean the whole function. Functions can be declared below where they're used. (And ideally as close to the root closure as possible – so they capture as little as possible.)

File: lib/class-factory.js [hunk 6]
```diff
@@ -177,13 +182,36 @@ function ClassFactory (vm) {
...
+        } finally {
+          setUsedClass(className, C);
+        }
       }
     }
 
     return new C(null);
   };
 
+  function getUsedClass(className) {
```
Comment (oleavr) [inline lib/class-factory.js]:
```suggestion
  function getUsedClass (className) {
```

Could you run this code through [semistandard](https://github.com/standard/semistandard)? Noticed missing space after method names (like this), but there might be other minor issues needing fixing.

File: lib/class-factory.js [hunk 7]
```diff
@@ -177,13 +182,36 @@ function ClassFactory (vm) {
...
+    let kclass;
```
Comment (oleavr) [inline lib/class-factory.js]:
```suggestion
    let klass;
```

^ Perhaps we should name it `klass` for consistency?

File: lib/class-factory.js [hunk 8]
```diff
@@ -177,13 +182,36 @@ function ClassFactory (vm) {
...
+      Thread.sleep(0.05);
+    }
+    if (kclass === undefined) {
+      classes[className] = PENDING_USE;
+    }
+    return kclass;
+  }
+
+  function setUsedClass(className, kclass) {
+    if (kclass !== undefined) {
```
Comment (oleavr) [inline lib/class-factory.js]: (Here as well.)

**Comment by oleavr** [comment]
Thanks!

**Pull request merged**